actors/migration/nv15/miner.go

```diff
@@ -3,79 +3,201 @@ package nv15
+			}
+
+			inSectors, err := miner6.LoadSectors(store, inRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+			}
+
+			prevOutSectors, err := miner7.LoadSectors(store, prevOutRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+			}
+
+			for _, change := range diffs {
+				switch change.Type {
+				case amt.Remove:
+					if err := prevOutSectors.Delete(change.Key); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+					}
+				case amt.Add:
+					fallthrough
+				case amt.Modify:
+					sectorNo := abi.SectorNumber(change.Key)
+					info, found, err := inSectors.Get(sectorNo)
+					if err != nil {
+						return cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+					}
+
+					if !found {
+						return cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+					}
+
+					if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info)); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+					}
+				}
+			}
+
+			outArray = prevOutSectors.Array
+		}
+
+		outRoot, err := outArray.Root()
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error writing new sectors AMT: %w", err)
+		}
+
+		if err = cache.Write(MinerPrevSectorsInKey(minerAddr), inRoot); err != nil {
+			return cid.Undef, xerrors.Errorf("failed to write previous sectors in key to cache: %w", err)
+		}
+
+		if err = cache.Write(MinerPrevSectorsOutKey(minerAddr), outRoot); err != nil {
+			return cid.Undef, xerrors.Errorf("failed to write previous sectors out key to cache: %w", err)
+		}
+
+		return outRoot, nil
+	})
+}
+func migrateDeadlines(ctx context.Context, store adt.Store, cache MigrationCache, deadlines cid.Cid, sectors cid.Cid) (cid.Cid, error) {
+	var inDeadlines miner6.Deadlines
+	err := store.Get(store.Context(), deadlines, &inDeadlines)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to read sectors array: %w", err)
+		return cid.Undef, err
 	}
 
-	outArray, err := adt.MakeEmptyArray(ctxStore, miner7.SectorsAmtBitwidth)
-	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to construct new sectors array: %w", err)
+	if miner6.WPoStPeriodDeadlines != miner7.WPoStPeriodDeadlines {
+		return cid.Undef, xerrors.Errorf("unexpected WPoStPeriodDeadlines changed from %d to %d",
+			miner6.WPoStPeriodDeadlines, miner7.WPoStPeriodDeadlines)
 	}
 
-	var sectorInfo miner6.SectorOnChainInfo
-	if err = inArray.ForEach(&sectorInfo, func(k int64) error {
-		return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
-	}); err != nil {
-		return cid.Undef, err
+	outDeadlines := miner7.Deadlines{Due: [miner7.WPoStPeriodDeadlines]cid.Cid{}}
+
+	for i, c := range inDeadlines.Due {
+		outDlCid, err := cache.Load(DeadlineKey(c), func() (cid.Cid, error) {
```

Comment:
This assumption is wrong. If the sectors CID has changed, we need to migrate the deadlines regardless.

We have two options:

1. Just migrate every time. Honestly, this is probably _fine_ as long as we're caching top-level actor migrations. If a miner is _active_, their deadlines will have changed significantly anyways. If a miner isn't active, the top-level actor cache will skip the miner.
2. Key by deadline CID _and_ sectors CID.

Reply:
Yessir, you right -- I think I'm gonna migrate every time, with a special case for the empty deadline and deadlines objects which I think will be a sizeable difference.